ARSIS-Unity/Assets/Scripts/EventManager/WebSocketClient.cs

```diff
@@ -0,0 +1,139 @@
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+
+namespace ARSIS.EventManager
+{
+    public class WebSocketClient
+    {
+        private Uri endpoint;
+        private CancellationTokenSource cancellation = new();
+        private ConcurrentQueue<string> sendQueue = new();
+        private ConcurrentQueue<string> receiveQueue = new();
+
+        public WebSocketClient(string endpoint) {
+            this.endpoint = new Uri(endpoint);
+        }
+
+        /// <summary>
+        /// Parses the string endpoint and provides a URI to the client.
+        /// </summary>
+        /// <param name="endpoint"></param>
+        public void SetEndpoint(string endpoint)
+        {
+            this.endpoint = new Uri(endpoint);
+        }
+
+        /// <summary>
+        /// Returns the endpoint.
+        /// </summary>
+        /// <returns>String of the endpoint.</returns>
+        public string GetEndpoint()
+        {
+            return endpoint.ToString();
+        }
+
+        /// <summary>
+        /// Gets an event from the receive queue of events from the WebSocket connection.
+        /// </summary>
+        /// <returns>BaseArsisEvent or null if queue is empty or failed to parse JSON.</returns>
+        public BaseArsisEvent GetEvent()
+        {
+            try
+            {
+                if (receiveQueue.TryDequeue(out string json))
+                    return PerformReflection(json);
+                return null;
+            }
+            catch (Exception e)
+            {
+                Debug.LogException(e);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Performs reflection on the JSON payload and returns an object with superclass BaseArsisEvent.
+        /// </summary>
+        /// <param name="json"></param>
+        /// <returns>BaseArsisEvent or null if failed to parse JSON.</returns>
+        private BaseArsisEvent PerformReflection(string json)
+        {
+            BaseArsisEvent unknownEvent = JsonUtility.FromJson<BaseArsisEvent>(json);
+            if (unknownEvent == null) return null;
+            if (unknownEvent.type.Equals(Telemetry.Type))

```

Comment:
I think here we can create a dictionary or do something that does this not by checking the type but by using generics to unjson it.